clean-data/clean-data.R

```diff
@@ -0,0 +1,700 @@
+        mutate_each(funs(grepl(findStr, ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    subDirty <- dirtyDat %>% filter(dirtyIdx) %>%
+        mutate_(.dots = setNames(list(varval), colName))
+
+    cleanDat <- dirtyDat %>% filter(!dirtyIdx) %>% bind_rows(subDirty)
+
+    cleanDat
+}
+
+# Clean Expected Earnings
+clean_expected_earnings <- function(cleanPart1) {
+    # Remove dollar signs from expected earnings
+    dollarIdx <- cleanPart1 %>% select(ExpectedEarning) %>%
+        mutate_each(funs(grepl("\\$", ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    dollarData <- cleanPart1 %>% filter(dollarIdx) %>%
+        mutate(ExpectedEarning = sub("\\$", "", ExpectedEarning))
+    cleanPart1 <- cleanPart1 %>% filter(!dollarIdx) %>% bind_rows(dollarData)
+
+    # Remove commas from expected earnings
+    commaIdx <- cleanPart1 %>% select(ExpectedEarning) %>%
+        mutate_each(funs(grepl(",", ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    commaData <- cleanPart1 %>% filter(commaIdx) %>%
+        mutate(ExpectedEarning = sub(",", "", ExpectedEarning))
+    cleanPart1 <- cleanPart1 %>% filter(!commaIdx) %>% bind_rows(commaData)
+
+    # Remove "k" from expected earnings
+    kIdx <- cleanPart1 %>% select(ExpectedEarning) %>%
+        mutate_each(funs(grepl("k", ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    kData <- cleanPart1 %>% filter(kIdx) %>%
+        mutate(ExpectedEarning = sub("k", "000", ExpectedEarning))
+    cleanPart1 <- cleanPart1 %>% filter(!kIdx) %>% bind_rows(kData)
+
+    # Change range of expected earnings into average
+    rangeIdx <- cleanPart1 %>% select(ExpectedEarning) %>%
+        mutate_each(funs(grepl("-", ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    rangeData <- cleanPart1 %>% filter(rangeIdx) %>%
+        mutate(ExpectedEarning = average_range_earning(ExpectedEarning))
+    cleanPart1 <- cleanPart1 %>% filter(!rangeIdx) %>% bind_rows(rangeData)
+
+    # Remove period from salaries like 50.000 which should be just 50000
+    thousandsIdx <- cleanPart1 %>% select(ExpectedEarning) %>%
+        mutate_each(funs(grepl("^\\d{2}\\.", ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    thousandsData <- cleanPart1 %>% filter(thousandsIdx) %>%
+        mutate(ExpectedEarning = sub("\\.", "", ExpectedEarning))
+    cleanPart1 <- cleanPart1 %>% filter(!thousandsIdx) %>%
+        bind_rows(thousandsData)
+
+    # Remove any non-numeric characters
+    numericIdx <- cleanPart1 %>% select(ExpectedEarning) %>%
+        mutate_each(funs(grepl("[A-Za-z]", ., ignore.case = TRUE))) %>%
+        unlist(use.names = FALSE)
+    numericData <- cleanPart1 %>% filter(numericIdx) %>%
+        mutate(ExpectedEarning = gsub("[A-Za-z']*", "", ExpectedEarning))
+    cleanPart1 <- cleanPart1 %>% filter(!numericIdx) %>%
+        bind_rows(numericData)
+
+    # Make all expected earnings less than 100 multiplied by 1000
```

Comment:
If we take the approach to guess the real intention, then here is a possible scheme for expected salary:

1. **```<19```** - cutoff (into blank) //too weird to me monthly and too small to be annual (because annual mostly used in rich countries, where salaries are high)
2. **```20-200```** - multiply by 1000 //too small to be monthly, but enough to be annual (I also remember to be confused for a few seconds should I put 5000 or 5K or 60 or 60K or 60000 or 60,000 or 60.0000 to fit the culture of your country)
3. **```201-499```** - cutoff (into blank) //too high for annual, too small for monthly for most countries
4. **```500-5999```** - multiply by 12 //this looks like monthly salary for poor and middle-rich countries
5. **```6000-200000```** - keep as it is //$6k as a minimum annual salary was picked intuitively because it's too big too be monthly in poor countries and in rich countries people use annual representation
6. **```>200000```** - set to 200000 //$200K maximum wage taken from stackoverflow survey

Finally, if we take this approach, we can make even more sophisticated system and normalize numbers according to the country of a respondent. Result would be accurate data, but the impact on average value might be very small.